Service Node Deregister Part 5

Doy-lee opened the pull request:
Internal review so I can get some feedback and get some of these TODOs addressed. This will be the first of probably a few reviews before being reviewed on dev because a lot of "temporary" code is still present in the commits.

I squashed all my commits and resplit them out into parts abit to help comprehension. I put deregister and partial deregister implementations into 1 commit because splitting the two proved to be abit difficult.

Any code that is prefixed with xx__* is generally temporary code and will be deleted before production.

The main flow to follow the code paths are,
```

1.a. simplewallet::xx__deregister_service_node(): Make a full cmd
2.a. wallet2::commit_tx(): To send the full deregistration to daemon RPC
3.a. core_rpc_server::on_send_raw_tx(): Submit full TX to core for validation
... And so forth, this follows the same code path as regular TX's.
```

```
1.b. simplewallet::xx__deregister_service_node(): Make a partial cmd
2.b. wallet2::commit_deregister_vote(): To send a partial deregistration vote to daemon RPC
3.b. core_rpc_server::on_submit_deregister_vote(): Submit vote to core for validation
4.b. core::add_deregister_vote(): Passes vote down to vote pool
5.b. loki::deregister_vote_pool::add_vote(): Validate the given vote, add it to the pool
From here, if the vote was valid all votes are collected, the full deregistration is constructed in 4b.
```

Commit:
Retrieve quorum list from height, reviewed

Commit:
Retrieve quorum list from height, reviewed

Commit:
Setup data structures for de/register TX

Commit:
Setup data structures for de/register TX

Commit:
Submit and validate partial/full deregisters

Commit:
Submit and validate partial/full deregisters

Commit:
Add P2P relaying of partial deregistration votes

Commit:
Add P2P relaying of partial deregistration votes

jcktm commented:
abit is not a word.

Doy-lee commented on src/cryptonote_core/service_node_deregister.cpp:
```diff
@@ -0,0 +1,734 @@
+  };
+
+  static crypto::hash hash_region_to_buf(void *buf, size_t buf_size, copy_region const *regions, int num_regions)
+  {
```
Hey I know. This is super unnecessary. After refactoring I forgot to merge this function down to just make_hash_from.

Doy-lee replied:
Fixed.

Doy-lee commented on src/cryptonote_core/service_node_deregister.h:
```diff
@@ -0,0 +1,130 @@
+  class deregister_vote_pool
+  {
+    public:
+      class pool_entry
```
Forgot to move these back to private scope.

Doy-lee reviewed (commented):
Some things I noticed after re-reviewing my changes.

Doy-lee commented on src/cryptonote_core/service_node_deregister.cpp:
```diff
@@ -731,4 +795,32 @@ namespace loki
+    for (auto it = m_deregisters.begin(); it != m_deregisters.end();)
+    {
+      time_t lifetime = now - it->time_group_created;
+      if (it->block_height < minimum_height && lifetime >= ALIVE_SECONDS_WINDOW)
```
Thoughts? I have votes living in the pool for 20 blocks but also keep them around if they haven't been alive for at least (20 blocks * target solve time) seconds. 

So if hash rate drops crazy and blocks mined slows down, we don't throw away votes. If blocks get mined quick but the votes haven't lived long enough they stick around so we can reuse the data on reorg (i.e. timing attacks)

Commit:
Code review adjustments for deregistration part 1

 - Fix check_tx_semantic
 - Remove signature_pod as votes are now stored as blobs. Serialization
   overrides don't intefere with crypto::signature anymore.

Commit:
Code review adjustments for deregistration part 1

 - Fix check_tx_semantic
 - Remove signature_pod as votes are now stored as blobs. Serialization
   overrides don't intefere with crypto::signature anymore.

jcktm commented on src/wallet/wallet2.cpp:
```diff
@@ -4581,7 +4581,29 @@ crypto::hash wallet2::get_payment_id(const pending_tx &ptx) const
+    THROW_WALLET_EXCEPTION_IF(!r, error::no_connection_to_daemon, "submit_deregister_vote");
+    THROW_WALLET_EXCEPTION_IF(resp.status == CORE_RPC_STATUS_BUSY, error::daemon_busy, "submit_deregister_vote");
+    THROW_WALLET_EXCEPTION_IF(resp.status != CORE_RPC_STATUS_OK, error::vote_rejected, resp.status, resp.reason);
+  }
```
Looks fine

jcktm commented on tests/core_proxy/core_proxy.h:
```diff
@@ -103,5 +104,9 @@ namespace tests
     bool fluffy_blocks_enabled() const { return false; }
     uint64_t prevalidate_block_hashes(uint64_t height, const std::list<crypto::hash> &hashes) { return 0; }
+
+    // TODO(loki): Write tests
```
This belongs on Monday

mizuki-hikaru reviewed (commented):
Pretty much everything looks fine to me, when the quorum stuff is integrated with the service node list I'll do another pass

Commit:
deregistration_vote_pool - changed sign/verify interface and removed repeated code

Commit:
deregistration_vote_pool - changed sign/verify interface and removed repeated code

Commit:
Misc review, fix sign/verify api, vote threshold

Commit:
Misc review, fix sign/verify api, vote threshold

Doy-lee commented on src/cryptonote_core/cryptonote_core.cpp:
```diff
@@ -1690,6 +1687,21 @@ namespace cryptonote
+        vvc.m_invalid_block_height = true;
+        return false;
+      }
+    }
```
Reject votes being passed around that are too old.

Commit:
Deregister/tx edge case handling for combinatoric votes

Commit:
Deregister/tx edge case handling for combinatoric votes

Doy-lee commented on src/cryptonote_core/blockchain.cpp:
```diff
@@ -2981,6 +2982,48 @@ bool Blockchain::check_tx_inputs(transaction& tx, tx_verification_context &tvc,
       return false;
     }
   }
+  else if (tx.version == transaction::version_3_deregister_tx)
```
Reject other versions of deregister TX's by checking the blockchain DB up to (VOTE_LIFETIME_BY_HEIGHT * 2) number of blocks from the height that the vote was submitted for.

Doy-lee commented on src/cryptonote_core/tx_pool.cpp:
```diff
@@ -106,6 +106,36 @@ namespace cryptonote
 
+  }
+  //---------------------------------------------------------------------------------
+  bool tx_memory_pool::have_deregister_tx_already(transaction const &tx) const
```
Reject TX's passed around with a different combination of votes that already are waiting in the mempool.

jcktm replied:
Do you also check previous blocks?

Doy-lee reviewed (commented):
@jcktm Please review. My understanding is now that if multiple daemons decide to take a block with a deregister TX (with different arrangement of votes) then eventually the network will choose the most successful chain and the other daemons will eventually drop their block and the deregister stored in the successful chain will become the canonical deregister TX.

The basic premise of implementing this is searching upwards around 60 blocks from the height the vote was submitted for TX V3s that have the matching block height and service node index they're voting on.

jcktm commented on src/cryptonote_core/cryptonote_core.cpp:
```diff
@@ -873,6 +875,44 @@ namespace cryptonote
+          return true;
+        });
+
+        if (!xx__is_service_node_registered(deregister.block_height, deregister.service_node_index))
```
Ok

jcktm replied:
As discussed, move to blockchain.cpp check_tx_inputs

Commit:
Store service node lists for the duration of deregister lifetimes

Commit:
Quorum min/max bug, sort node list, fix node to test list

Doy-lee commented on src/cryptonote_core/blockchain.cpp:
```diff
@@ -2981,6 +2982,48 @@ bool Blockchain::check_tx_inputs(transaction& tx, tx_verification_context &tvc,
+    }
+
+    const uint64_t height            = deregister.block_height;
+    const size_t num_blocks_to_check = loki::service_node_deregister::VOTE_LIFETIME_BY_HEIGHT * 2;
```
Possibly. Votes stay around for VOTE_LIFETIME_BY_HEIGHT && VOTE_LIFETIME_BY_HEIGHT * TARGET_DIFFICULTY_V2 (so time in seconds since the vote was created), so votes may be included in blocks after the vote lifetime depending on if blocks are mined real quick. Maybe 2x the maximum is overkill.

jcktm replied:
Remove timestamp checks, make it height only.

Add height check to check_tx_inputs, verify that check_tx_inputs only works on transactions that are being considered for the next block.

Doy-lee replied:
Gonna need clarification on that height check.

jcktm replied:
> so votes may be included in blocks after the vote lifetime depending on if blocks are mined real quick

Please make it so that votes are limited by exactly vote lifetime blocks, no more, even if blocks are mined quickly. 2h expected time is more than enough.

Commit:
core, service_node_list: separated address from service node pubkey

Commit:
Change quorum to store acc pub address, fix oob bug

Commit:
Merge branch 'GenerateServiceNodeListsForHeights' of github.com:doy-lee/loki into ServiceNodeDeregisterPart1

Doy-lee commented:
Part 3 @jcktm, the last 3 commits is the range you want to review.
https://github.com/loki-project/loki/pull/89/files/5733a7e65ad0e4181b89ab0084cfbaa82f3e245b..82f5dda03588ff2601d1e3c8c1e6b86362d19c90

jcktm commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -95,40 +95,45 @@ namespace service_nodes
-          return memcmp(reinterpret_cast<const void*>(&a), reinterpret_cast<const void*>(&b), sizeof(a)) < 0;
-        });
-    return ret;
+    std::vector<const cryptonote::account_public_address *> result;
```
const crypto::public_key * or crypto::public_key

Using a pointer here instead of the whole public_key object saves 11ms on my machine. But it also makes it easier for people to make hard to find memory corruption bugs, because the pointers will be invalidated when the quorum for that height drops off the cache. For you this may be clear now, but for someone else using this function it may not be so obvious.

This is similar to returning the quorum_state, that the correct way to do this efficiently would be to use a shared pointer.

I have a personal preference for just doing the copy with crypto::public_key, and then only implementing microoptimizations if speed here actually becomes a problem.

jcktm commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -95,40 +95,45 @@ namespace service_nodes
     for (size_t i = 0; i < circular_quorum_state_buffer::SIZE; i ++)
     {
-      quorum_state const *check = m_quorum_states.quorums + i;
+      const quorum_state *check = m_quorum_states.quorums + i;
```
When can the quorum_state become invalidated?

jcktm commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -348,11 +371,30 @@ namespace service_nodes
+        {
+          crypto::public_key& service_node_key = m_service_nodes_keys[address];
+          m_rollback_events.push_back(std::unique_ptr<rollback_event>(new rollback_change(block_height, address, iter->second, service_node_key)));
+          m_service_nodes_last_reward.erase(iter);
```
Add m_service_nodes_keys.erase(address);

jcktm commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -348,11 +371,30 @@ namespace service_nodes
+        }
+        else
+        {
+          MWARNING("Tried to kick off a service node that is no longer registered: " << epee::string_tools::pod_to_hex(address.m_spend_public_key));
```
s/address/key/

this applies to all quorum stuff.

jcktm commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -507,23 +544,24 @@ namespace service_nodes
     quorum_state& quorum_state = m_quorum_states.next_free_slot();
```
You may be storing a quorum for a height that already exists in the case of reorgs. So you have to search your existing buffer and see if an entry for the height already exists.

This would be easier to read, faster to implement, and more efficient as a map.

jcktm commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -507,23 +544,24 @@ namespace service_nodes
     {
       quorum_state.height                     = height;
-      std::vector<crypto::public_key>& quorum = quorum_state.quorum_nodes;
+      std::vector<cryptonote::account_public_address>& quorum = quorum_state.quorum_nodes;
```
crypto::public_key

jcktm commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -542,8 +580,8 @@ namespace service_nodes
   {
-    int index            = this->circular_index++ % SIZE;
-    quorum_state& result = quorums[index];
+    this->circular_index  = (this->circular_index + 1) % SIZE;
```
Nice catch

jcktm commented on src/cryptonote_basic/tx_extra.h:
```diff
@@ -179,16 +181,61 @@ namespace cryptonote
+    struct vote
+    {
+      signature_pod signature;
+      uint32_t      voters_quorum_index;
```
Y'know, or not

jcktm commented on src/cryptonote_core/cryptonote_core.cpp:
```diff
@@ -996,6 +1036,30 @@ namespace cryptonote
+      }
+
+      std::vector<crypto::public_key> quorum;
+      if (!get_quorum_list_for_height(deregister.block_height, quorum))
```
Okay that's been finalized now =]

jcktm commented on src/cryptonote_core/cryptonote_core.cpp:
```diff
@@ -1561,6 +1651,103 @@ namespace cryptonote
+    // Generate index mapping to pub_keys
+    std::vector<size_t> pub_keys_indexes;
+    {
+      pub_keys_indexes.reserve(pub_keys.size());
```
All good, no problem with this, was just curious

jcktm commented on src/cryptonote_core/cryptonote_core.cpp:
```diff
@@ -1561,6 +1651,103 @@ namespace cryptonote
+    return true;
+  }
+  //-----------------------------------------------------------------------------------------------
+  bool core::get_quorum_list_for_height(uint64_t height, std::vector<crypto::public_key>& quorum) const
```
> then we can modify this at that point.

jcktm commented on src/cryptonote_core/service_node_deregister.cpp:
```diff
@@ -0,0 +1,825 @@
+    return result;
+  }
+
+  crypto::hash service_node_deregister::make_unsigned_vote_hash(const cryptonote::tx_extra_service_node_deregister& deregister)
```
Fixed

jcktm commented on src/cryptonote_core/blockchain.cpp:
```diff
@@ -3002,6 +3003,48 @@ bool Blockchain::check_tx_inputs(transaction& tx, tx_verification_context &tvc,
+      for (blobdata const &blob : txs)
+      {
+        transaction existing_tx;
+        if (!parse_and_validate_tx_from_blob(blob, existing_tx))
```
Print some warnings for these continues

jcktm commented on src/cryptonote_core/cryptonote_core.cpp:
```diff
@@ -1218,8 +1309,14 @@ namespace cryptonote
+
+      if(update_miner_blocktemplate)
+         update_miner_block_template();
+    }
```
All good

Doy-lee replied:
I'm going to move this into blockchain::handle_block_to_main_chain() because at the end of the function there's a call to m_tx_pool.on_blockchain_inc(..) which we sort of agreed that this was likely where the txpool is meant to flush out tx's that are now invalid because of the new block (but isn't at the moment). So similarly we can flush out deregister votes in the pool in the same spot with the added bonus we also have access to the TX's so I don't need to re-parse the block data again.

It means adding to the blockchain constructor a reference to the deregister vote pool though and whatever else it takes to get that compiling (blackball etc).

jcktm replied:
Fair enough, let's just get this finished

Doy-lee commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -207,6 +212,23 @@ namespace service_nodes
     return money_transferred >= m_blockchain.get_staking_requirement(block_height);
   }
 
+  bool service_node_list::process_deregistration_tx(const cryptonote::transaction& tx, cryptonote::account_public_address& address)
```
Done

Doy-lee reviewed (commented):
@jcktm

Commit:
Code review for expiring votes, acc keys to pub_key, improve err msgs

Doy-lee commented:
@jcktm Please review.

Commit:
Merge branch 'dev' into ServiceNodeDeregisterPart1

Commit:
Add early out for is_deregistration_tx and protect against quorum changes

jcktm commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -518,19 +535,14 @@ namespace service_nodes
 
       // Shuffle indexes
-      if (0) // TODO(doyle): Temp. For debugging with deterministic lists
+      if (0) // xx__remove_me For debugging with deterministic lists
```
Please remove

Doy-lee replied:
Removed.

jcktm commented on src/simplewallet/simplewallet.cpp:
```diff
@@ -4719,7 +4719,7 @@ bool simple_wallet::stake_all(const std::vector<std::string> &args_)
 
   uint64_t unlock_block = 0;
-  uint64_t locked_blocks = STAKING_REQUIREMENT_LOCK_BLOCKS;
+  uint64_t locked_blocks = STAKING_REQUIREMENT_LOCK_BLOCKS + 20; /* DO NOT COMMIT: Because the time you submit and the time the block gets accepted is always going to be > 1, which fails the is_reg_tx check */
```
Why do not commit? This buffer needs to be included in cryptonote_config.h:

     52 #define STAKING_REQUIREMENT_LOCK_BLOCKS                 (30*24*31)                                                                                                                                          | 723         tie(mixRing, index) = populateFromBlockchain(inPk, mixin);
     53 #define STAKING_REQUIREMENT_LOCK_BLOCKS_EXCESS          20                                                                                                                                                  | 724         return genRct(message, inSk, destinations, amounts, mixRing, amount_keys, kLRki, msout, index, outSk, false, hwdev);
     54 #define STAKING_RELOCK_WINDOW_BLOCKS                    (30*6)

Doy-lee replied:
Updated.

jcktm commented on src/simplewallet/simplewallet.cpp:
```diff
@@ -4853,137 +4853,90 @@ bool simple_wallet::xx__deregister_service_node(const std::vector<std::string> &
-
-  tools::wallet2::pending_tx ptx;
-  ptx.tx.unlock_time = 0;
+  // xx__deregister_node <partial|full> <height> <node id>
```
remove this function please.

Doy-lee replied:
All debug code has been removed.

mizuki-hikaru reviewed (changes requested):
Deregister part 5, the saga continues

Doy-lee commented on src/cryptonote_core/service_node_deregister.cpp:
```diff
@@ -675,7 +685,7 @@ namespace loki
   bool deregister_vote_pool::add_vote(const service_node_deregister::vote& new_vote,
                                       cryptonote::vote_verification_context& vvc,
-                                      const service_nodes::quorum_state& quorum_state,
+                                      service_nodes::quorum_state &quorum_state,
```
Mistake, added it back.,

Doy-lee commented on src/cryptonote_core/service_node_deregister.cpp:
```diff
@@ -743,6 +753,33 @@ namespace loki
+      desired_group.block_height       = deregister.block_height;
+      desired_group.service_node_index = deregister.service_node_index;
+
+      const auto &group_it = m_deregisters.find(desired_group);
```
Updated.

Doy-lee commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -212,18 +202,30 @@ namespace service_nodes
     {
-      address = state->nodes_to_test[deregister.service_node_index];
-      return true;
+      const crypto::public_key &snode_pubkey = state->nodes_to_test[deregister.service_node_index];
```
This was updated in the subsequent pull request. You were probably reviewing a stale version of the page.

Doy-lee commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -408,6 +420,11 @@ namespace service_nodes
       m_rollback_events.pop_back();
     }
+
+    while (m_quorum_states.end()->first > height)
```
Updated.

Commit:
Remove debug code, fix segfault

Doy-lee reviewed (commented):
Part 5 it is, this removes debug code and code review adjustments. @jcktm

Doy-lee commented on src/cryptonote_core/service_node_deregister.cpp:
```diff
@@ -526,6 +527,15 @@ namespace loki
     crypto::hash hash = make_hash_from(block_height, service_node_index);
     for (auto& key_and_sig : keys_and_sigs)
     {
+      crypto::signature xx__bypass_sig;
```
Removed.

jcktm commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -326,6 +433,11 @@ namespace service_nodes
       m_rollback_events.pop_back();
     }
+
+    while (!m_quorum_states.empty() && (--m_quorum_states.end())->first > height)
```
>= height

jcktm commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -326,6 +433,11 @@ namespace service_nodes
       m_rollback_events.pop_back();
     }
+
+    while (!m_quorum_states.empty() && (--m_quorum_states.end())->first > height)
```
`>= height`

Doy-lee replied:
Fixed up

jcktm commented on src/cryptonote_core/tx_pool.cpp:
```diff
@@ -224,6 +303,16 @@ namespace cryptonote
+      // earliest one kept. Eventually other daemons will sync this block which
+      // will become the canonical deregister TX.
+      if (tx.version == transaction::version_3_deregister_tx)
+        kept_by_block = false;
```
Please remove

Doy-lee replied:
Removed.

mizuki-hikaru reviewed (approved):
![](https://media.giphy.com/media/13zeE9qQNC5IKk/giphy.gif)

Commit:
Remove irrelevant check for tx v3 in blockchain, fix >= height for pruning quorum states

Incorrect assumption that a transaction can be kept in the chain if it could
eventually become invalid, because if it were the chain would be split and
eventually these transaction would be dropped. But also that we should not
override the pre-existing logic which handles this case anyway.
